DOC-4807 Key visualizer (#16479)
cockroachdb/docs: v23.1/ui-key-visualizer.md

The key visualizer is a useful troubleshooting tool when experiencing performance problems with your cluster, surfacing historical and current KV [hot spots](performance-best-practices-overview.html#hot-spots) in your keyspace, drawing attention to [range splits](architecture/distribution-layer.html#range-splits), and highlighting potentially-unnecessary [full-table scans](make-queries-fast.html) that might benefit from the creation of a targeted index, among others.

The key visualizer is disabled by default. Once [enabled](#enable-the-key-visualizer), the key visualizer continuously collects keyspace usage data across your cluster in the background at a [configurable sampling rate](#key-visualizer-customization). Data shown in the key visualizer is retained for a maximum period of seven days.

To access the key visualizer, [navigate to the DB Console](ui-overview.html#db-console-access) and click **Advanced Debug** in the left-hand navigation, then click **Key Visualizer** under the **Even More Advanced Debugging** heading.

## Required privileges

> To view the **Advanced Debug** page, and work with the key visualizer, the user must be a member of the `admin` role or must have the `VIEWDEBUG` [system privilege](security-reference/authorization.html#supported-privileges) defined.

Reviewer: ```suggestion
To view the **Advanced Debug** page and work with the key visualizer, the user must be a member of the `admin` role or must have the `VIEWDEBUG` [system privilege](security-reference/authorization.html#supported-privileges) defined.
```

## Enable the key visualizer

> To use the key visualizer, the `keyvisualizer.enabled` [cluster setting](cluster-settings.html) must be set to `true`, using the [`SET CLUSTER SETTING`](set-cluster-setting.html) statement:

Reviewer: ```suggestion
To use the key visualizer, the [cluster setting](cluster-settings.html) `keyvisualizer.enabled` must be set to `true`, using the [`SET CLUSTER SETTING`](set-cluster-setting.html) statement:
```

Author: We do in fact tend to present cluster settings in the way I've presented this one here:

- ```the `.*` \[cluster setting\]``` :: 794 results
- ```the \[cluster setting\](.*) `.*```` : 71 results.

Neither is wrong, neither is perfect, but one is more common.

Reviewer: Oh, thanks for checking! I'll open a style issue - I prefer the latter so that people don't have to get through a setting name to know what we're talking about, in some cases.

Author: Remember, in the future it **could** look like this!

- ```the [`keyvisualizer.enabled`](cluster-settings.html#setting-keyvisualizer_enabled) cluster setting.```

## Key visualizer features

Once you have enabled the key visualizer, CockroachDB will begin monitoring keyspace usage across your cluster. Data is only collected while the key visualizer is enabled; you cannot visualize data from a time period before you enabled it.

When navigating to the **Key Visualizer** page in the DB Console, after a brief loading time, CockroachDB will present the collected data in a visualization designed to help you see data traffic trends at a glance.

<img src="{{ 'images/v23.1/key-visualizer-hot-range.png' | relative_url }}" alt="Key visualizer showing hot range" style="max-width:80%" />

The key visualizer presents the following information:

- The entire [table data keyspace](architecture/distribution-layer.html#table-data-kv-structure) is represented on the Y-axis, and is broken up into buckets representing whole ranges of the keyspace, or aggregated ranges if the number of ranges in the keyspace exceeds the [configured number of buckets](#key-visualizer-customization).

- Time is represented on the X-axis, with its granularity (i.e., frequency of data collection) being controlled by the [configured sample period](#key-visualizer-customization).

> - Keyspace activity is visualized on a color scale from black to red, representing "cold" and "hot" respectively. Thus, a range shown in bright red indicates a [hot spot](performance-best-practices-overview.html#hot-spots), while a range shown in black indicates a range with little to no active reads or writes. Hot spots are identified relative to other ranges, e.g., a range that receives one write a minute could be considered a hot spot if the rest of the ranges on the cluster aren't receiving any. A range shown in red is therefore not necessarily itself indicative of a problem, but a useful visual on the path to troubleshooting one.

Reviewer: I felt that the last sentence could be more clear—the last clause seems to assume (depending on interpretation) that there _is_ a problem to troubleshoot, but the first clause raises the possibility that this isn't the case. Here's just one potential way to make it more clear, if you agree with the message.

```suggestion
- Keyspace activity is visualized on a color scale from black to red, representing "cold" and "hot" respectively. Thus, a range shown in bright red indicates a [hot spot](performance-best-practices-overview.html#hot-spots), while a range shown in black indicates a range with little to no active reads or writes. Hot spots are identified relative to other ranges; for example, a range that receives one write per minute could be considered a hot spot if the rest of the ranges on the cluster aren't receiving any. A range shown in red is therefore not necessarily indicative of a problem, but it may call attention to an area that needs attention to prevent a problem.
```

Reviewer: > it may call attention to an area that needs attention

Oops on the word repetition! Not my best work, but hopefully the suggestion got my point across.

Author: Missed it in my initial review of your review. Thank you for reviewing your review! (caught and pushed a fix, should be ready to merge after this).

If you adjust either of these values, new keyspace activity data collected by the key visualizer will reflect the new values, but previously-collected activity data will remain unchanged.

## Troubleshooting performance with the key visualizer

When troubleshooting a performance issue with your cluster, use the key visualizer to identify which range or ranges in your keyspace to focus on. If needed, consider adjusting the key visualizer's [granularity of data collection](#key-visualizer-customization) of sample data collection to more closely focus on a problematic range of keys if necessary, especially if they keyspace is very large and you have the storage headroom in your cluster.

The key visualizer was designed to make potentially problematic ranges stand out visually; as such, bright red spot are generally good places to begin a performance investigation. For example, consider the following cases:

### Identifying hot spots

> The following image shows the key visualizer highlighting a series of [hot spots](performance-best-practices-overview.html#hot-spots): ranges with much higher-than-average write rates comparative to the rest of the cluster.

Reviewer: ```suggestion
The following image shows the key visualizer highlighting a series of [hot spots](performance-best-practices-overview.html#hot-spots): ranges with much higher-than-average write rates as compared to the rest of the cluster.
```

<img src="{{ 'images/v23.1/key-visualizer-range-split.png' | relative_url }}" alt="Key visualizer showing range split" style="max-width:80%" />

> **Remediation:** If you've identified a potentially-problematic range as a hot spot, follow the recommended best practices to [reduce hot spots](performance-best-practices-overview.html#reduce-hot-spots). In the case of the screenshot above, the increased write cadence is due to a series of [range splits](architecture/distribution-layer.html#range-splits), where a range experiencing a large volume of incoming writes is splitting its keyspace to accommodate the growing range. This is often part of normal operation, but can be indicative of a data modeling issue if the range split is unexpected, or causing cluster performance issues.

Reviewer: ```suggestion
**Remediation:** If you've identified a potentially-problematic range as a hot spot, follow the recommended best practices to [reduce hot spots](performance-best-practices-overview.html#reduce-hot-spots). In the case of the screenshot above, the increased write cadence is due to a series of [range splits](architecture/distribution-layer.html#range-splits), where a range experiencing a large volume of incoming writes is splitting its keyspace to accommodate the growing range. This is often part of normal operation, but can be indicative of a data modeling issue if the range split is unexpected or causing cluster performance issues.
```
